### PR TITLE
simplify nnx_basics

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

* Simplify `nnx_basics`

### CODE REVIEW

**Review by IvyZX — APPROVED** [review]

File: docs_nnx/nnx_basics.ipynb [hunk 1]
```diff
@@ -384,21 +384,21 @@
...
-    "4. [`State`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/state.html#flax.nnx.State) updates for the [`nnx.BatchNorm`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/normalization.html#flax.nnx.BatchNorm) and [`nnx.Dropout`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/stochastic.html#flax.nnx.Dropout) layers are automatically propagated by [`nnx.scan`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/transforms.html#flax.nnx.scan)."
+    "1. The custom `create_model` function takes in a key and returns an `MLP` object, since you create five keys and use `nnx.vmap` over `create_model` a stack of 5 `MLP` objects is created.\n",
+    "2. The `nnx.scan` is used to iteratively apply each `MLP` in the stack to the input `x`.\n",
+    "3. The nnx.scan (consciously) deviates from `jax.lax.scan` and instead mimics nnx.vmap, which is more expressive. nnx.scan allows specifying multiple inputs, the scan axes of each input/output, and the position of the carry.\n",
+    "4. `State` updates for the `BatchNorm` and `Dropout` layers are automatically propagated by nnx.scan."
```
Comment (8bitmp3) [inline docs_nnx/nnx_basics.ipynb:401]: We'll have user experience SEO confusion if we don't specify if common terms / APIs names that are used in normal ML English like State/BatchNorm/Dropout are Flax NNX APIs or Linen (old docs are in search results / used in codebase).

File: docs_nnx/nnx_basics.ipynb [hunk 2]
```diff
@@ -384,21 +384,21 @@
    "source": [
     "There are two things happening in this example that are worth mentioning:\n",
     "\n",
-    "1. The updates to each of the [`nnx.BatchNorm`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/normalization.html#flax.nnx.BatchNorm) and [`nnx.Dropout`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/stochastic.html#flax.nnx.Dropout) layer's state is automatically propagated from within `loss_fn` to `train_step` all the way to the `model` reference outside.\n",
-    "2. The `optimizer` holds a mutable reference to the `model` - this relationship is preserved inside the `train_step` function making it possible to update the model's parameters using the optimizer alone.\n",
+    "1. The updates to each of the `BatchNorm` and `Dropout` layer's state is automatically propagated from within `loss_fn` to `train_step` all the way to the `model` reference outside.\n",
```
Comment (8bitmp3) [inline docs_nnx/nnx_basics.ipynb:387]: Outside of the smaller NNX community, external users would need to know if BatchNorm / Dropout is an Flax NNX or flax.nn (Linen) BatchNorm. We should at least specify this here (entire doc). cc @IvyZX

File: docs_nnx/nnx_basics.ipynb [hunk 3]
```diff
@@ -332,11 +332,11 @@
...
     "\n",
-    "In the following example, you define a `train_step` function that takes a `MLP` model, an [`nnx.Optimizer`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/training/optimizer.html#module-flax.nnx.optimizer), and a batch of data, and returns the loss for that step. The loss and the gradients are computed using the [`nnx.value_and_grad`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/transforms.html#flax.nnx.value_and_grad) transform over the `loss_fn`. The gradients are passed to the optimizer's [`nnx.Optimizer.update`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/training/optimizer.html#flax.nnx.optimizer.Optimizer.update) method to update the `model`'s parameters."
+    "In the following example, you define a `train_step` function that takes a `MLP` model, an `Optimizer`, and a batch of data, and returns the loss for that step. The loss and the gradients are computed using the `nnx.value_and_grad` transform over the `loss_fn`. The gradients are passed to the optimizer's `update` method to update the model's parameters."
```
Comment (8bitmp3) [inline docs_nnx/nnx_basics.ipynb:339]: The `update` method here is the Flax `nnx.Optimizer.update` method, and `update` is a common English term and a common method name in ML frameworks. We need to make it clear that it belongs to Flax NNX `Optimizer`, or new users will find it difficult to learn that it's a reference to `flax.nnx.optimizer.Optimizer.update`. The word "update" is used all over ML and RL. Removing this is a UX and readability downgrade. @IvyZX.

Comment (8bitmp3) [reply docs_nnx/nnx_basics.ipynb:339]: @jburnim WDYT

File: docs_nnx/nnx_basics.ipynb (hunk 2 above)
Comment (8bitmp3) [reply docs_nnx/nnx_basics.ipynb:387]: @jburnim WDYT

File: docs_nnx/nnx_basics.ipynb [hunk 4]
```diff
@@ -41,9 +41,9 @@
...
     "\n",
-    "Let's begin by creating a `Linear` [`nnx.Module`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/module.html). As shown next, dynamic state is usually stored in `nnx.Param`s, and static state (all types not handled by NNX) such as integers or strings are stored directly. Attributes of type `jax.Array` and `numpy.ndarray` are also treated as dynamic states, although storing them inside `nnx.Variable`s, such as `Param`, is preferred. Also the [`nnx.Rngs`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/rnglib.html#flax.nnx.Rngs) object can be used to get new unique keys based on a root PRNG key passed to the constructor."
+    "Let's begin by creating a Linear `Module`. As shown next, dynamic state is usually stored in `Param`s, and static state (all types not handled by NNX) such as integers or strings are stored directly. Attributes of type `jax.Array` and `numpy.ndarray` are also treated as dynamic states, although storing them inside Variables, such as Param, is preferred. Also the `Rngs` object can be used to get new unique keys based on a root PRNG key passed to the constructor."
```
Comment (8bitmp3) [inline docs_nnx/nnx_basics.ipynb:46]: Flax NNX `Variable` (`flax.nnx.Variable) may be confused for just ML variables (English), hence the original backticks. Same goes for `nnx.Param`s. We can't just remove them here and in `Param`s while leaving the capitalization (e.g. Params and Variables) - they are common English names, so this can create confusion for new NNX users @jburnim @IvyZX 

`nnx.Rngs` needs to be NNX `Rngs` - by removing the reference to NNX you can create a confusion becayse RNGs and PRNGs exist in JAX outside of Flax NNX API (@jakevdp ).

File: docs_nnx/nnx_basics.ipynb [hunk 5]
```diff
@@ -657,10 +657,10 @@
...
     "\n",
-    "To address this, the Flax NNX API has [`nnx.split`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/graph.html#flax.nnx.split), which allows you to pass one or more [`nnx.filterlib.Filter`s](https://flax.readthedocs.io/en/latest/guides/filters_guide.html) to partition the [`nnx.Variable`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/variables.html#flax.nnx.Variable)s into mutually exclusive [`nnx.State`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/state.html#flax.nnx.State)s. Flax NNX uses `Filter` create `State` groups in APIs (such as [`nnx.split`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/graph.html#flax.nnx.split), [`nnx.state()`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/graph.html#flax.nnx.state), and many of NNX transforms).\n",
+    "To address this, the Flax NNX API has `nnx.split`, which allows you to pass one or more `Filter`s to partition the `Variable`s into mutually exclusive `State`s. Flax NNx uses `Filter` create `State` groups in APIs (such as `nnx.split`, `nnx.state`, and many of NNX transforms).\n",
```
Comment (8bitmp3) [inline docs_nnx/nnx_basics.ipynb:663]: If we don't mention that `Filter` is a Flax NNX `filterlib` `Filter` (i.e. `flax.nnx.filterlib.Filter`), users may find it very difficult to find the reference `Filter` API in our codebase or API docs.

In addition, "`Filter`" (like "`State`", or "`split`") is a common English word, and users may try to do a search and filter for "`Filter`" ... We should specify it's a Flax NNX `filterlib` `Filter`. Naming ia hard. cc @jburnim @jakevdp @IvyZX

**Comment by cgarciae** [comment]
@8bitmp3 with this updates I'm optimizing for aesthetics, current guide has too many links which makes it unpleasant to read through. I don't think linking to every symbol does us favors here, I'd consider it if there was a way to make it more invisible but currently links stand out too much.

**Comment by 8bitmp3** [comment]
Ok SGTM @cgarciae Let's work on auto-linking so users don't have to use the search bar a lot.

For namespaces etc of APIs, we must still use full names to avoid name confusion. e.g. "Filter" (like "State", or "split") is a common English word, and users may try to do a search and filter for "Filter" ... We should specify it's a Flax NNX filterlib Filter. Perhaps, in another PR.
